core/jobs/playtime_entries.py

```diff
@@ -0,0 +1,116 @@
+import csv
+import os
+from collections import defaultdict
+from datetime import date, timedelta
+from tempfile import TemporaryFile
+
+from sqlalchemy.sql.functions import sum
+
+from core.config import Configuration
+from core.model import get_one
+from core.model.edition import Edition
+from core.model.time_tracking import IdentifierPlaytime, IdentifierPlaytimeEntry
+from core.util.datetime_helpers import utc_now
+from core.util.email import EmailManager
+from scripts import Script
+
+
+class PlaytimeEntriesSummationScript(Script):
+    def do_run(self):
+        # Collect everything from one hour ago, reducing entries still in flux
+        cuttoff = utc_now() - timedelta(hours=1)
+
+        # Reap older processed entries
+        deleted = (
+            self._db.query(IdentifierPlaytimeEntry)
+            .filter(
+                IdentifierPlaytimeEntry.processed == True,
+            )
+            .delete()
+        )
+        self.log.info(f"Deleted {deleted} old entries.")
+
+        # Fetch the unprocessed entries
+        result = self._db.query(IdentifierPlaytimeEntry).filter(
+            IdentifierPlaytimeEntry.processed == False,
+            IdentifierPlaytimeEntry.timestamp <= cuttoff,
+        )
+        by_identifier = defaultdict(int)
+
+        # Aggregate entries per identifier-timestamp tuple
+        # Since timestamps should be on minute-boundaries the aggregation
+        # can be written to IdentifierPlaytimes directly
+        for entry in result.all():
+            by_identifier[
+                (entry.identifier, entry.timestamp)
+            ] += entry.total_seconds_played
+            entry.processed = True
+
+        for id_ts, seconds in by_identifier.items():
+            identifier, timestamp = id_ts
+            playtime = IdentifierPlaytime.add(identifier, timestamp, seconds)
+            self.log.info(
+                f"Added {seconds} to {identifier.urn} for {timestamp}: new total {playtime.total_seconds_played}."
+            )
+
+        self._db.commit()
+
+
+class PlaytimeEntriesEmailReportsScript(Script):
+    def do_run(self):
+        """Send a quarterly report with aggregated playtimes via email"""
+        # 3 months prior, shifted to the 1st of the month
+        cutoff = utc_now() - timedelta(days=90)
+        cutoff = cutoff.replace(day=1).date()
+        # Until the 1st of the current month
+        until = date.today().replace(day=1)
```

Comment:
`date.today()` returns the date for local -- rather than UTC -- time, but I think we need UTC time here.

Reply:
Oh right, silly